misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt

```diff
@@ -0,0 +1,149 @@
+import java.sql.Connection
+import java.util.Collections
+import java.util.NavigableMap
+import java.util.NavigableSet
+import java.util.TreeMap
+import java.util.TreeSet
+import java.util.regex.Pattern
+import javax.persistence.PersistenceException
+
+private val logger = getLogger<SchemaMigrator>()
+
+/**
+ * Manages **available** and **applied** schema migrations.
+ *
+ * Available schema migrations are SQL files in the datasource's `migrations_path` directory. Each
+ * file should contain SQL statements terminated by a `;`. The files should be named like
+ * `v100__exemplar.sql` with a `v`, an integer version, two underscores, a description, and the
+ * `.sql` suffix. The integer identifier is the migration version. Versions do not need to be
+ * sequential. They are applied in increasing order.
+ *
+ * Applied schema migrations are tracked in the database in a `schema_version` table. Migrations may
+ * be applied either by [SchemaMigrator.applyAll] or manually. When you applying schema changes
+ * manually you must add a row to the `schema_version` table to record which version was applied.
+ */
+
+internal class SchemaMigrator(
+  private val resourceLoader: ResourceLoader,
+  private val sessionFactory: SessionFactory,
+  private val config: DataSourceConfig
+) {
+  /** Returns a map from version to path. */
+  fun availableMigrations(): NavigableMap<Int, String> {
+    val resources = resourceLoader.list(config.migrations_path)
+    return TreeMap(resources.associateBy { resourceVersion(it) })
+  }
+
+  /** Creates the `schema_version` table if it does not exist. Returns the applied migrations. */
+  fun initialize(): NavigableSet<Int> {
+    try {
+      val result = appliedMigrations()
+      logger.info {
+        "${config.database} has ${result.size} migrations applied;" +
+            " latest is ${result.lastOrNull()}"
+      }
+      return result
+    } catch (e: PersistenceException) {
+      sessionFactory.doWork {
+        val statement = createStatement()
+        statement.addBatch("""
+            |CREATE TABLE schema_version (
+            |  version varchar(50) NOT NULL,
+            |  installed_by varchar(30) DEFAULT NULL,
+            |  UNIQUE KEY (version)
+            |);
+            |""".trimMargin())
+        statement.executeBatch()
+      }
+      logger.info {
+        "${config.database} has no migrations applied; created the schema_version table"
+      }
+      return Collections.emptyNavigableSet()
+    }
+  }
+
+  /**
+   * Returns the versions of applied migrations. Throws a [PersistenceException] if the migrations
+   * table has not been initialized.
+   */
+  fun appliedMigrations(): NavigableSet<Int> {
+    sessionFactory.openSession().use { session ->
+      @Suppress("UNCHECKED_CAST") // createNativeQuery returns a raw Query.
+      val query = session.createNativeQuery("SELECT version FROM schema_version") as Query<String>
+      return TreeSet(query.list().map { it.toInt() })
```

Comment:
here you create a List then a TreeSet, but you could mapTo the TreeSet directly
```
return query.list().mapTo(TreeSet()) { it.toInt() }
```
I'm not picky though

Reply:
I think it does the same thing underneath?